File: qbopt/analysis/intervals.py
"""Where each value is live, as ranges rather than as sets per block.

`allocate.live()` answers live-in and live-out per block, which is enough
to say whether two values ever overlap and not enough for anything the
allocator wants to do about it. Splitting a range needs to know where it
begins; spilling well needs to know how long it is; coalescing two values
needs to know that their ranges meet at exactly one point.

LLVM's shape, and the names are its: a **slot index** numbers every point a
value can start or stop being live, a **segment** is a half-open run of
them, and a **live interval** is the segments one value occupies.
`llvm/lib/CodeGen/LiveIntervals.cpp` is the original.

Two slots per instruction rather than LLVM's four. It uses Slot_Block,
Slot_EarlyClobber, Slot_Register and Slot_Dead to separate an operand read
before the write from one written before the read; nothing here yet needs
early-clobber, and inventing the distinction before a pass asks for it
would be four times the indices for none of the answers.
"""

from dataclasses import dataclass
from dataclasses import replace

from qbopt.model import lir
from qbopt.analysis import loops as loopy

# Points per instruction: where it reads, and where it writes. A value read
# at `use` and one written at `def` of the same instruction do not overlap,
# which is what makes `add ax,bx` not interfere ax with itself.
USE = 0
DEF = 1
PER_INSN = 2


@dataclass(frozen=True, slots=True)
class Segment:
    """A half-open run of slot indices one value occupies."""

    start: int
    end: int

    def overlaps(self, other: "Segment") -> bool:
        return self.start < other.end and other.start < self.end


@dataclass(frozen=True, slots=True)
class Interval:
    """Every segment one value occupies, and what spilling it would cost."""

    value: int
    segments: tuple[Segment, ...]
    weight: float = 0.0

    @property
    def size(self) -> int:
        """How many slots this value is live for, which is not its span.

        A value live at the top and bottom of a body and dead in between
        occupies two short segments, and spilling it frees a register for
        everything between them. Measuring the span instead would say it
        frees nothing.
        """
        return sum(one.end - one.start for one in self.segments)

    def overlaps(self, other: "Interval") -> bool:
        mine, theirs = iter(self.segments), iter(other.segments)
        one, two = next(mine, None), next(theirs, None)
        while one is not None and two is not None:
            if one.overlaps(two):
                return True
            if one.end <= two.end:
                one = next(mine, None)
            else:
                two = next(theirs, None)
        return False


@dataclass(frozen=True, slots=True)
class Indexes:
    """Every instruction's slot number, and every block's span."""

    at: dict[int, int]  # id(insn) -> the instruction's first slot
    span: dict[int, tuple[int, int]]  # block address -> [first, last)
    order: tuple[int, ...]  # block addresses, in the order they are numbered


def indexed(body: lir.LirBody) -> Indexes:
    """Number every point a value can start or stop being live.

    In block order, which is the order they are emitted in. LLVM numbers in
    layout order for the same reason: a range that looks contiguous has to
    be contiguous in the thing that runs.
    """
    at: dict[int, int] = {}
    span: dict[int, tuple[int, int]] = {}
    next_slot = 0
    for block in body.blocks:
        first = next_slot
        # A phi's result is defined before the block's first instruction, on
        # the edge rather than in the block. One slot, so it can be live
        # out of the predecessor and in here without the two touching.
        next_slot += PER_INSN
        for one in block.insns:
            at[id(one)] = next_slot
            next_slot += PER_INSN
        span[block.at] = (first, next_slot)
    return Indexes(at, span, tuple(block.at for block in body.blocks))


def intervals(body: lir.LirBody, index: Indexes | None = None) -> dict[int, Interval]:
    """The live interval of every value in this body, weighted.

    Built from block liveness and then refined inside each block, which is
    how LiveIntervalCalc does it: the set says which values cross the
    block's edges, and walking the instructions backwards says where inside
    it each one actually starts and stops.
    """
    index = index or indexed(body)
    ranges = _ranges(body, index)
    weight = _weights(body, index, ranges)
    return {value: replace(one, weight=weight.get(value, 0.0)) for value, one in ranges.items()}


def _ranges(body: lir.LirBody, index: Indexes) -> dict[int, Interval]:
    """Where each value is live, before anything prices it."""
    from qbopt.backend import allocate

    live_in, live_out = allocate.live(body)
    pieces: dict[int, list[Segment]] = {}
    for block in body.blocks:
        first, last = index.span[block.at]
        alive: dict[int, int] = {one: last for one in live_out[block.at]}
        written: set[int] = set()
        for one in reversed(block.insns):
            slot = index.at[id(one)]
            for value in one.defines:
                written.add(value)
                pieces.setdefault(value, []).append(Segment(slot + DEF, alive.pop(value, slot + DEF + 1)))
            for value in one.uses:
                alive.setdefault(value, slot + USE + 1)
        # A phi's result is defined at the top of the block, on the edge
        # rather than by any instruction in it.
        for phi in block.phis:
            written.add(phi.result)
            pieces.setdefault(phi.result, []).append(Segment(first + DEF, alive.pop(phi.result, first + DEF + 1)))
        # Whatever is still alive arrived from a predecessor and was live
        # from the block's first slot.
        for value, end in alive.items():
            if end > first:
                pieces.setdefault(value, []).append(Segment(first, end))
        # Live through: in at the top, out at the bottom, untouched between.
        for value in live_in[block.at]:
            if value not in written and value not in alive:
                pieces.setdefault(value, []).append(Segment(first, last))
    return {value: Interval(value, tuple(_merged(runs))) for value, runs in pieces.items()}


def _merged(runs: list[Segment]) -> list[Segment]:
    """Overlapping or touching segments joined, in order."""
    out: list[Segment] = []
    for one in sorted(runs, key=lambda x: (x.start, x.end)):
        if out and one.start <= out[-1].end:
            out[-1] = Segment(out[-1].start, max(out[-1].end, one.end))
            continue
        out.append(one)
    return out


def depths(body: lir.LirBody) -> dict[int, int]:
    """How deeply each block is nested in loops."""
    out = {block.at: 0 for block in body.blocks}
    for loop in loopy.loops(list(body.blocks), body.entry):
        for at in loop.body:
            if at in out:
                out[at] += 1
    return out


# What one reference costs per level of loop nesting. LLVM asks
# BlockFrequencyInfo for the block's frequency relative to the entry, which
# with no profile is this estimate: ten iterations assumed per loop.
PER_LEVEL = 10

# Added to the size before dividing, so that a very short interval's weight
# is dominated by how often it is referenced rather than by an accident of
# where the slots fell. LLVM's `25 * InstrDist`, in this module's slots.
GRACE = 25 * PER_INSN


def weights(body: lir.LirBody, index: Indexes | None = None) -> dict[int, float]:
    """What spilling each value would cost. See `_weights` for the formula."""
    index = index or indexed(body)
    return _weights(body, index, _ranges(body, index))


def _weights(body: lir.LirBody, index: Indexes, ranges: dict[int, Interval]) -> dict[int, float]:
    """`references weighted by loop depth / (live slots + grace)`.

    LLVM's normalizeSpillWeight. The division is the part a plain count
    misses: two values referenced equally often are not equally worth
    keeping if one is live for three instructions and the other for the
    whole body -- spilling the long one frees a register for longer, so it
    is the cheaper one to spill and its weight has to say so.
    """
    deep = depths(body)
    total: dict[int, float] = {}
    for block in body.blocks:
        each = float(PER_LEVEL ** deep.get(block.at, 0))
        for one in block.insns:
            for value in (*one.defines, *one.uses):
                total[value] = total.get(value, 0.0) + each
    return {value: found / (ranges[value].size + GRACE if value in ranges else GRACE) for value, found in total.items()}

File: crates/llrm-core/src/analysis/intervals.rs
//! Port of `qbopt/analysis/intervals.py`: where each value is live, as
//! ranges rather than as sets per block.
//!
//! LLVM's `LiveIntervals`, two slots per instruction: where it reads, and
//! where it writes.

use std::collections::BTreeSet;
use std::sync::Arc;

use crate::support::hash::{IndexMap, IndexSet};

use crate::analysis::loops as loopy;
use crate::backend::allocate;
use crate::model::lir::{Insn, LirBlock, LirBody};
use crate::model::mir::MirBlock;

// Points per instruction: where it reads, and where it writes.
pub const USE: i64 = 0;
pub const DEF: i64 = 1;
pub const PER_INSN: i64 = 2;

/// A half-open run of slot indices one value occupies.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub struct Segment {
    pub start: i64,
    pub end: i64,
}

impl Segment {
    pub fn overlaps(&self, other: &Segment) -> bool {
        self.start < other.end && other.start < self.end
    }
}

/// Every segment one value occupies, and what spilling it would cost.
#[derive(Clone, Debug, PartialEq)]
pub struct Interval {
    pub value: u32,
    pub segments: Vec<Segment>,
    pub weight: f64,
}

impl Interval {
    pub fn new(value: u32, segments: Vec<Segment>) -> Self {
        Self { value, segments, weight: 0.0 }
    }

    /// How many slots this value is live for, which is not its span.
    pub fn size(&self) -> i64 {
        self.segments.iter().map(|one| one.end - one.start).sum()
    }

    pub fn overlaps(&self, other: &Interval) -> bool {
        let (mut mine, mut theirs) = (self.segments.iter(), other.segments.iter());
        let (mut one, mut two) = (mine.next(), theirs.next());
        while let (Some(first), Some(second)) = (one, two) {
            if first.overlaps(second) {
                return true;
            }
            if first.end <= second.end {
                one = mine.next();
            } else {
                two = theirs.next();
            }
        }
        false
    }
}

/// Python `id(insn)`: the instruction's identity in the body that holds it.
pub fn key(one: &Arc<Insn>) -> usize {
    Arc::as_ptr(one) as usize
}

/// Every instruction's slot number, and every block's span.
#[derive(Clone, Debug)]
pub struct Indexes {
    pub at: IndexMap<usize, i64>, // id(insn) -> the instruction's first slot
    pub span: IndexMap<i64, (i64, i64)>, // block address -> [first, last)
    pub order: Vec<i64>,          // block addresses, in the order they are numbered
}

impl Indexes {
    /// The slot of the instruction at `position` in `block`, or the block's end past its last.
    pub fn slot(&self, block: &LirBlock, position: usize) -> i64 {
        block.insns.get(position).map_or(self.span[&block.at].1, |one| self.at[&key(one)])
    }

    /// The position in `block` of the instruction whose slots hold `slot`: -1
    /// for the block's entry, its length for its end.
    pub fn position(&self, block: &LirBlock, slot: i64) -> i64 {
        let (first, last) = self.span[&block.at];
        if slot < first + PER_INSN {
            return -1;
        }
        if slot >= last {
            return block.insns.len() as i64;
        }
        block.insns.iter().rposition(|one| self.at[&key(one)] <= slot).map_or(-1, |at| at as i64)
    }
}

/// Number every point a value can start or stop being live.
pub fn indexed(body: &LirBody) -> Indexes {
    let mut at = IndexMap::default();
    let mut span = IndexMap::default();
    let mut next_slot = 0;
    for block in &body.blocks {
        let first = next_slot;
        // A phi's result is defined before the block's first instruction.
        next_slot += PER_INSN;
        for one in &block.insns {
            at.insert(key(one), next_slot);
            // A meta instruction takes no slot, as LLVM's SlotIndexes skip
            // debug instructions: a range across one is no longer.
            if !one.is_meta() {
                next_slot += PER_INSN;
            }
        }
        span.insert(block.at, (first, next_slot));
    }
    Indexes { at, span, order: body.blocks.iter().map(|block| block.at).collect() }
}

/// The live interval of every value in this body, weighted.
pub fn intervals(body: &LirBody, index: Option<&Indexes>) -> IndexMap<u32, Interval> {
    let owned;
    let index = match index {
        Some(index) => index,
        None => {
            owned = indexed(body);
            &owned
        }
    };
    let ranges = _ranges(body, index);
    let weight = _weights(body, index, &ranges);
    ranges
        .into_iter()
        .map(|(value, one)| {
            let weight = weight.get(&value).copied().unwrap_or(0.0);
            (value, Interval { weight, ..one })
        })
        .collect()
}

/// The `group` run ending at `position`: where it starts.
fn _group_start(block: &LirBlock, position: usize) -> usize {
    let one = &block.insns[position];
    let mut first = position;
    if one.group.is_some() {
        while first > 0 && block.insns[first - 1].group == one.group {
            first -= 1;
        }
    }
    first
}

/// Where each value is live, before anything prices it.
///
/// Python builds `pieces` by iterating sets; only the map's order differs,
/// and nothing reads it in order.
fn _ranges(body: &LirBody, index: &Indexes) -> IndexMap<u32, Interval> {
    let (live_in, live_out) = allocate::live(body);
    let mut pieces: IndexMap<u32, Vec<Segment>> = IndexMap::default();
    for block in &body.blocks {
        let (first, last) = index.span[&block.at];
        let mut alive: IndexMap<u32, i64> = live_out[&block.at].iter().map(|one| (*one, last)).collect();
        let mut written: BTreeSet<u32> = BTreeSet::new();
        let mut position = block.insns.len() as i64 - 1;
        while position >= 0 {
            let at = position as usize;
            let one = &block.insns[at];
            let first_in_group = _group_start(block, at);
            let group = &block.insns[first_in_group..=at];
            let slot = if one.group.is_some() {
                index.at[&key(group.last().expect("a group holds its last"))]
            } else {
                index.at[&key(one)]
            };
            let boundary = slot + DEF;
            let defined: IndexSet<u32> = group.iter().flat_map(|item| item.defines.iter().copied()).collect();
            for value in defined {
                written.insert(value);
                let end = alive.shift_remove(&value).unwrap_or(boundary + 1);
                pieces.entry(value).or_default().push(Segment { start: boundary, end });
            }
            let used: IndexSet<u32> = group.iter().flat_map(|item| item.uses.iter().copied()).collect();
            for value in used {
                alive.entry(value).or_insert(boundary);
            }
            position = first_in_group as i64 - 1;
        }
        // A phi's result is defined at the top of the block.
        for phi in &block.phis {
            written.insert(phi.result);
            let end = alive.shift_remove(&phi.result).unwrap_or(first + DEF + 1);
            pieces.entry(phi.result).or_default().push(Segment { start: first + DEF, end });
        }
        // Whatever is still alive arrived from a predecessor.
        for (value, end) in &alive {
            if *end > first {
                pieces.entry(*value).or_default().push(Segment { start: first, end: *end });
            }
        }
        // Live through: in at the top, out at the bottom, untouched between.
        for value in &live_in[&block.at] {
            if !written.contains(value) && !alive.contains_key(value) {
                pieces.entry(*value).or_default().push(Segment { start: first, end: last });
            }
        }
    }
    pieces
        .into_iter()
        .map(|(value, runs)| (value, Interval::new(value, _merged(runs))))
        .collect()
}

/// Join overlapping segments, retaining touching definition boundaries.
pub fn _merged(mut runs: Vec<Segment>) -> Vec<Segment> {
    runs.sort_by_key(|x| (x.start, x.end));
    let mut out: Vec<Segment> = Vec::new();
    for one in runs {
        if let Some(last) = out.last_mut() {
            if one.start < last.end {
                *last = Segment { start: last.start, end: last.end.max(one.end) };
                continue;
            }
        }
        out.push(one);
    }
    out
}

/// `loopy.loops` reads only `at` and `succ`; MIR blocks carry both.
pub fn _graph(blocks: &[LirBlock]) -> Vec<MirBlock> {
    blocks
        .iter()
        .map(|block| MirBlock::new(block.at, Vec::new(), Vec::new(), block.succ.clone()))
        .collect()
}

/// How deeply each block is nested in loops.
pub fn depths(body: &LirBody) -> IndexMap<i64, u32> {
    let mut out: IndexMap<i64, u32> = body.blocks.iter().map(|block| (block.at, 0)).collect();
    for found in loopy::loops(&_graph(&body.blocks), Some(body.entry)) {
        for at in &found.body {
            if let Some(depth) = out.get_mut(at) {
                *depth += 1;
            }
        }
    }
    out
}

/// What one reference costs per level of loop nesting.
pub const PER_LEVEL: i64 = 10;

/// Added to the size before dividing. LLVM's `25 * InstrDist`.
pub const GRACE: i64 = 25 * PER_INSN;

/// `float(PER_LEVEL ** depth)`.
pub fn level(depth: u32) -> f64 {
    PER_LEVEL.pow(depth) as f64
}

/// What spilling each value would cost. See `_weights` for the formula.
pub fn weights(body: &LirBody, index: Option<&Indexes>) -> IndexMap<u32, f64> {
    let owned;
    let index = match index {
        Some(index) => index,
        None => {
            owned = indexed(body);
            &owned
        }
    };
    _weights(body, index, &_ranges(body, index))
}

/// `references weighted by loop depth / (live slots + grace)`.
fn _weights(body: &LirBody, _index: &Indexes, ranges: &IndexMap<u32, Interval>) -> IndexMap<u32, f64> {
    let deep = depths(body);
    let mut total: IndexMap<u32, f64> = IndexMap::default();
    for block in &body.blocks {
        let each = level(deep.get(&block.at).copied().unwrap_or(0));
        for one in &block.insns {
            for value in one.defines.iter().chain(&one.uses) {
                *total.entry(*value).or_insert(0.0) += each;
            }
        }
    }
    total
        .into_iter()
        .map(|(value, found)| {
            let size = match ranges.get(&value) {
                Some(one) => one.size() + GRACE,
                None => GRACE,
            };
            (value, found / size as f64)
        })
        .collect()
}
